`upgrade-provider` does not update `sdk/go.mod` pulumi version

### What happened?

During a routine upgrade, I kept getting "no required module" errors. Running the upgrade manually, I isolated the error to the build go sdk step.

### Expected Behavior

Successful build of the go SDK

### Steps to reproduce

Set [this sample sdk.go.mod's pulumi version](https://github.com/pulumi/pulumi-artifactory/blob/main/sdk/go.mod#L7) to a laughably low Pulumi version, such as `v3.36.0`

Run `make build_go`

Observe errors such as

```
go/artifactory/accessToken.go:13:2: no required module provides package github.com/pulumi/pulumi/sdk/v3/go/pulumix; to add it:
	go get github.com/pulumi/pulumi/sdk/v3/go/pulumix
go/artifactory/accessToken.go:13:2: no required module provides package github.com/pulumi/pulumi/sdk/v3/go/pulumix; to add it:
	go get github.com/pulumi/pulumi/sdk/v3/go/pulumix
make: *** [build_go] Error 1
```

### Output of `pulumi about`

```
CLI          
Version      3.81.0
Go Version   go1.21.1
Go Compiler  gc
```

### Additional context

I believe not auto-updating the `sdk/go.mod` was an oversight in the tool that we should correct.

We should always update the SDK's pulumi version to latest as well.

It is, unfortunately, [missing from the full playbook as well.](https://github.com/pulumi/platform-providers-team/blob/main/playbooks/tf-provider-updating.md)

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/upgrade/steps-helpers.go
+++ b/upgrade/steps-helpers.go
@@ -76,6 +76,35 @@ func originalGoVersionOf(ctx context.Context, repo ProviderRepo, file, needleMod
 	return module.Version{}, false, nil
 }
 
+// Look up the version of the go dependency requirement of a given module in a given modfile.
+func currentGoVersionOf(modFile, lookupModule string) (module.Version, bool, error) {
+	fileData, err := os.ReadFile(modFile)
+	if err != nil {
+		return module.Version{}, false, err
+	}
+	goMod, err := modfile.Parse(modFile, fileData, nil)
+	if err != nil {
+		return module.Version{}, false, fmt.Errorf("%s: %w",
+			modFile, err)
+	}
+
+	// We can only look up requirements in this way.
+	for _, replacement := range goMod.Replace {
+		if replacement.Old.Path == lookupModule {
+			return module.Version{}, false, fmt.Errorf(
+				"module %s is being replaced, cannot lookup version", lookupModule,
+			)
+		}
+	}
+
+	for _, requirement := range goMod.Require {
+		if requirement.Mod.Path == lookupModule {
+			return requirement.Mod, true, nil
+		}
+	}
+	return module.Version{}, false, nil
+}
+
 func baseFileAt(ctx context.Context, repo ProviderRepo, file string) ([]byte, error) {
 	cmd := exec.CommandContext(ctx, "git", "show", repo.defaultBranch+":"+file)
 	cmd.Dir = repo.root

--- a/upgrade/steps-helpers.go
+++ b/upgrade/steps-helpers.go
@@ -121,6 +121,10 @@ func prBody(ctx Context, repo ProviderRepo,
 		fmt.Fprintf(b, "- Upgrading pulumi-terraform-bridge/pf from %s to %s.\n",
 			goMod.Pf.Version, targetBridge)
 	}
+
+	if ctx.UpgradeSdkVersion {
+		fmt.Fprintf(b, "- Upgrading pulumi SDK to `latest`.\n")
+	}
 	if parts := strings.Split(tfSDKUpgrade, " -> "); len(parts) == 2 {
 		fmt.Fprintf(b, "- Upgrading pulumi/terraform-plugin-sdk from %s to %s.\n",
 			parts[0], parts[1])

--- a/upgrade/upgrade-provider.go
+++ b/upgrade/upgrade-provider.go
@@ -293,6 +293,8 @@ func UpgradeProvider(ctx Context, repoOrg, repoName string) error {
 		repo.workingBranch = "upgrade-code-migration"
 	} else if ctx.UpgradePfVersion {
 		repo.workingBranch = fmt.Sprintf("upgrade-pf-version-to-%s", targetPfVersion)
+	} else if ctx.UpgradeSdkVersion {
+		repo.workingBranch = "upgrade-pulumi-sdk"
 	} else {
 		return fmt.Errorf("calculating branch name: unknown action")
 	}
@@ -343,6 +345,28 @@ func UpgradeProvider(ctx Context, repoOrg, repoName string) error {
 				return step.Cmd(exec.CommandContext(ctx, "go", "mod", "tidy")).
 					In(repo.providerDir())
 			}))
+
+		// Now that we've updated the bridge version, we need to update the corresponding pulumi version in sdk/go.mod.
+		// It needs to match the version used in provider/go.mod, which is *not* necessarily `latest`.
+		var newSdkVersion string
+		getNewPulumiVersionStep := step.F("Get Pulumi SDK version", func() (string, error) {
+			modFile := filepath.Join(repo.root, "provider", "go.mod")
+			lookupModule := "github.com/pulumi/pulumi/sdk/v3"
+			pulumiMod, found, err := currentGoVersionOf(modFile, lookupModule)
+			if err != nil {
+				return "", err
+			}
+			if !found {
+				return "", fmt.Errorf("%s: %s not found\n", modFile, lookupModule)
+			}
+			return pulumiMod.Version, nil
+
+		}).AssignTo(&newSdkVersion)
+
+		updateToBridgePulumiVersionStep := step.Computed(func() step.Step {
+			return step.Cmd(exec.CommandContext(ctx, "go", "get", "github.com/pulumi/pulumi/sdk/v3@"+newSdkVersion)).In(repo.sdkDir())
+		})
+		steps = append(steps, getNewPulumiVersionStep, updateToBridgePulumiVersionStep)
 
 	}
 	if ctx.UpgradeSdkVersion {
@@ -396,6 +420,7 @@ func UpgradeProvider(ctx Context, repoOrg, repoName string) error {
 	artifacts := append(steps,
 		step.Cmd(exec.CommandContext(ctx, "go", "mod", "tidy")).In(repo.providerDir()),
 		step.Cmd(exec.CommandContext(ctx, "go", "mod", "tidy")).In(repo.examplesDir()),
+		step.Cmd(exec.CommandContext(ctx, "go", "mod", "tidy")).In(repo.sdkDir()),
 		step.Computed(func() step.Step {
 			if ctx.RemovePlugins {
 				return step.Cmd(exec.CommandContext(ctx,

